Add and Delete display names via GitOps

jkatz01 opened the pull request:
<!-- Add the related story/sub-task/bug number, like Resolves #123, or remove if NA -->
**Related issue:** Resolves #36509 and #36513
-  batch set in-house apps will now add or delete display names with
-  batch set software installers will now delete display names (could already add)

# Checklist for submitter

If some of the following don't apply, delete the relevant line.

## Testing

- [x] Added/updated automated tests
- [ ] Where appropriate, [automated tests simulate multiple hosts and test for host isolation](https://github.com/fleetdm/fleet/blob/main/docs/Contributing/reference/patterns-backend.md#unit-testing) (updates to one hosts's records do not affect another)

- [x] QA'd all new/changed functionality manually

jkatz01 committed:
Add display names in batch set in-house apps

jkatz01 committed:
wip: delete display names

jkatz01 committed:
Delete display names in batch set installers

jkatz01 committed:
Merge branch 'main' into 36509-ipa-gitops-display-names

jkatz01 committed:
Factor out titleID

jkatz01 committed:
Add test for batch in-house apps

jkatz01 committed:
Add test for batch software installers

jkatz01 committed:
Delete display names when removing all software, ignore empty display names

jkatz01 committed:
Merge branch 'main' into 36509-ipa-gitops-display-names

iansltx reviewed (commented):
IMO there are a few tweaks worth making on how we delete/re-add names here. Also worth checking if VPP display names had a similar ordering issue to software packages. Probably not, but...

Also, what was the underlying issue that caused display names to be out of order for software packages? Guessing you were able to repro that/figured out the cause?

jkatz01 commented on server/datastore/mysql/in_house_apps.go:
```diff
@@ -1118,18 +1144,31 @@ VALUES
 			return ctxerr.Wrap(ctx, err, "delete obsolete in-house installers")
 		}
 
+		if _, err := tx.ExecContext(ctx, deleteInHouseAppDisplayNames, globalOrTeamID); err != nil {
```
I think this is an oversight that happened because I wanted this part to remove the display names that were swapped to the wrong titles, so I made it completely wipe everything every time. 
Would we want an approach like `stdn.software_title_id NOT IN (?)`, `titleIDs` similar to a few other statements in that function? 

There's also another issue I can think of right now which is for installers with an empty display names in the list, we need to check if there is an existing display name to update, or if nothing should be done.

jkatz01 commented on server/datastore/mysql/in_house_apps.go:
```diff
@@ -979,6 +987,20 @@ VALUES
 	%s
 `
 
+	const deleteInHouseAppDisplayNames = `
```
> Do we need an AND NOT EXISTS covering in house apps?

Probably not, depending on what we decide I should do about the second comment here.

> Also, does anything need to be tweaked on the VPP apps side to make sure VPP app setting doesn't
clear display names specific to in-house apps?

I can check that VPP apps don't destroy anything, but on a first glance at the code I can't really tell where display names are being updated.

iansltx replied:
Worth checking through code paths to confirm where they're being updated I think, in case e.g. they're running via additional API calls (which would be brittle).

jkatz01 committed:
wip: change display name add/delete logic

jkatz01 committed:
Change how adding/deleting names works for software installers

jkatz01 committed:
Change how adding/deleting names works for in house apps

iansltx committed:
Merge branch 'main' into 36509-ipa-gitops-display-names

iansltx commented on server/datastore/mysql/in_house_apps.go:
```diff
@@ -979,6 +987,46 @@ VALUES
+		AND vat.global_or_team_id = stdn.team_id);
+`
+
+	const deleteAllInHouseDisplayNames = `
```
If we throw an "and not exists" query for IPAs in here rather than going by title ID, we can use the same one everywhere.

iansltx commented on server/datastore/mysql/software_installers_test.go:
```diff
@@ -3737,6 +3737,11 @@ func testSoftwareTitleDisplayName(t *testing.T, ds *Datastore) {
+		DumpTable(t, q, "software_installers")
+		DumpTable(t, q, "software_title_display_names")
+		return nil
+	})
```
Need to delete this

iansltx reviewed (commented):
Sorry, got one more sorta architectural bit of feedback here. We're gettin' there...

jkatz01 committed:
Feedback changes
Use a simpler query to delete names not in list and reuse it to delete all names
Add small comments explaining reuse and deletions
Small test fix

jkatz01 committed:
very small formatting change

iansltx committed:
Add test assertions to ensure batch updates don't affect other installer types

iansltx reviewed (approved):
Added some test assertions to cover what I mentioned on Thu or Fri, so we're good to go here. There's more work to do for VPP apps, but that's got its own ticket so this can go in and get cherry-picked.
